service/serializer/src/main/java/org/jolokia/service/serializer/GSonSerializer.java

```diff
@@ -0,0 +1,201 @@
+                json = JsonPath.parse(json, jsonPathconfiguration).read(path).toString();
+            }
+            return wrapString(json);
+        } catch (PathNotFoundException e) {
+            getJolokiaContext().error(String.format("Unexpected error when serializing %s", pValue.getClass().getSimpleName()), e);
+            return null;
+        } catch (InvalidJsonException e) {
+            getJolokiaContext().error(String.format("Unexpected error when serializing %s", pValue.getClass().getSimpleName()), e);
+            return null;
+        } catch (JsonIOException e) {
+            getJolokiaContext().error(String.format("Unexpected error when serializing %s", pValue.getClass().getSimpleName()), e);
+            return null;
+        }
+    }
+
+    public Object deserialize(String pExpectedClassName, Object pValue) {
+        if (pValue == null) {
+            return null;
+        }
+        Class<?> expectedClass = ClassUtil.classForName(pExpectedClassName);
+        if (expectedClass == null) {
+            return null;
+        }
+        if (pValue instanceof String) {
+            if (((String) pValue).isEmpty() && expectedClass == String.class) {
+                return pValue;
+            }
+        }
+        try {
+            Gson gson = getGson();
+            return gson.fromJson(toJSON(pValue).toJSONString(), expectedClass);
+        } catch (JsonSyntaxException e) {
+            getJolokiaContext().error(String.format("Unexpected error when deserializing %s", pValue), e);
+            return null;
+        }
+    }
+
+    private Gson getGson() {
+        return new GsonBuilder()
+                .serializeNulls()
+                .serializeSpecialFloatingPointValues()
+                .create();
+    }
+
+    public WriteRequestValues setInnerValue(Object pOuterObject, Object pNewValue, List<String> pPathParts) throws AttributeNotFoundException, IllegalAccessException, InvocationTargetException {
+        if (pOuterObject == null) {
+            return null;
+        }
+        String originalJson = ((JSONAware) serialize(pOuterObject, Collections.<String>emptyList(), null)).toJSONString();
+        JsonElement newValue = (JsonElement) deserialize(JsonElement.class.getName(), pNewValue);
+        JSONAware originalValue = (JSONAware) serialize(pOuterObject, pPathParts, null);
+
+        String updatedJson;
+        if (originalValue == null) {
+            String lastPathElement = pPathParts.remove(pPathParts.size() - 1);
+            String path = buildPath(pPathParts);
+            updatedJson = JsonPath.parse(originalJson, jsonPathconfiguration).put(path, lastPathElement, newValue).jsonString();
+        } else {
+            String path = buildPath(pPathParts);
+            updatedJson = JsonPath.parse(originalJson, jsonPathconfiguration).set(path, newValue).jsonString();
+        }
+        return new WriteRequestValues(deserialize(pOuterObject.getClass().getName(), updatedJson), originalValue);
+    }
```

Comment:
The important point here is that `pOuterObject` is updated inline. It's not about getting an updated JSON structure but the object *itself* needs to be updated. That is what the JMX write operation is for, you update an JMX attribute. In that case not the attribute itself but at a deeper level (this is a feature of Jolokia). Afaiu do you only upate a JSON representation of the object which you then return.

It probably arguable whether this feature really belongs into a serializer. Maybe we should move this `setInnerValue()` completely out of this interface and make it an intrinsic feature of Jolokia. I remember that I tried that sometimes ago, but failed. 

The more I think about it, this method really doesn't belong here as it does have to do with serialization only marginally. I'll try to move this out of the serializer, though not sure wheter I suceed. 

Any idea welcome here ;-)